cbor/record.go

```diff
@@ -0,0 +1,206 @@
+package cbor
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/ipfs/go-cid"
+	"github.com/ipfs/go-ipld-cbor"
+	"github.com/ipfs/go-ipld-format"
+	ic "github.com/libp2p/go-libp2p-core/crypto"
+	mh "github.com/multiformats/go-multihash"
+	"github.com/textileio/go-textile-core/crypto"
+	"github.com/textileio/go-textile-core/thread"
+	pb "github.com/textileio/go-textile-threads/pb"
+)
+
+func init() {
+	cbornode.RegisterCborType(record{})
+}
+
+type record struct {
+	Block cid.Cid
+	Sig   []byte
+	Prev  cid.Cid `refmt:",omitempty"`
+}
+
+func NewRecord(ctx context.Context, dag format.DAGService, block format.Node, prev cid.Cid, sk ic.PrivKey, key crypto.EncryptionKey) (thread.Record, error) {
+	payload := block.RawData()
+	if prev.Defined() {
+		payload = append(payload, prev.Bytes()...)
+	}
+	sig, err := sk.Sign(payload)
+	if err != nil {
+		return nil, err
+	}
+	obj := &record{
+		Block: block.Cid(),
+		Sig:   sig,
+		Prev:  prev,
+	}
+	node, err := cbornode.WrapObject(obj, mh.SHA2_256, -1)
+	if err != nil {
+		return nil, err
+	}
+	coded, err := EncodeBlock(node, key)
+	if err != nil {
+		return nil, err
+	}
+
+	err = dag.AddMany(ctx, []format.Node{coded})
+	if err != nil {
+		return nil, err
+	}
+
+	return &Record{
+		Node:  coded,
+		obj:   obj,
+		block: block,
+	}, nil
+}
+
+func GetRecord(ctx context.Context, dag format.DAGService, id cid.Cid, key crypto.DecryptionKey) (thread.Record, error) {
+	coded, err := dag.Get(ctx, id)
+	if err != nil {
+		return nil, err
+	}
+	return RecordFromNode(coded, key)
+}
+
+func RecordFromNode(coded format.Node, key crypto.DecryptionKey) (thread.Record, error) {
+	obj := new(record)
+	node, err := DecodeBlock(coded, key)
+	if err != nil {
+		return nil, err
+	}
+	err = cbornode.DecodeInto(node.RawData(), obj)
+	if err != nil {
+		return nil, err
+	}
+	return &Record{
+		Node: coded,
+		obj:  obj,
+	}, nil
+}
+
+func RecordToProto(ctx context.Context, dag format.DAGService, rec thread.Record) (*pb.Record, error) {
```

Comment:
When sending records to other peers, we serialized the node and all it's children (minus any links in the body). This means we don't have to worry about handling a bunch of pins... just send the whole thing.